### PR TITLE
ref: Rationalize Kafka topic config

### PR DESCRIPTION
**Pull request by lynnagara**
This brings the concept of logical vs physical topic override to Sentry, which is fairly closely aligned with how topic management is done in Snuba.

Both the old way (KAFKA_TOPICS) and new way (KAFKA_TOPIC_OVERRIDES and KAFKA_TOPIC_TO_CLUSTER) will be supported until we have fully migrated over in production and all client code in this repository no longer reference the old variables.

### CODE REVIEW

File: src/sentry/conf/types/topic_definition.py [hunk 1]
```diff
@@ -1,7 +1,39 @@
-from __future__ import annotations
-
+from enum import Enum
```
Comment (fpacifici) [inline src/sentry/conf/types/topic_definition.py:1]: Shall we call this kafka_topic_definition ?

File: src/sentry/conf/types/topic_definition.py [hunk 2]
```diff
@@ -1,7 +1,39 @@
...
 from typing import TypedDict
 
 
+class Topic(Enum):
```
Comment (fpacifici) [inline src/sentry/conf/types/topic_definition.py:5]: Nice I like this.

File: src/sentry/conf/server.py [hunk 3]
```diff
@@ -3458,16 +3458,17 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
...
-    "events": KAFKA_EVENTS_SUBSCRIPTIONS_RESULTS,
-    "transactions": KAFKA_TRANSACTIONS_SUBSCRIPTIONS_RESULTS,
-    "generic-metrics": KAFKA_GENERIC_METRICS_SUBSCRIPTIONS_RESULTS,
-    "sessions": KAFKA_SESSIONS_SUBSCRIPTIONS_RESULTS,
-    "metrics": KAFKA_METRICS_SUBSCRIPTIONS_RESULTS,
+# Mapping of default Kafka topic name to custom names
+KAFKA_TOPIC_OVERRIDES: Mapping[str, str] = {}
+
+# Mapping of default Kafka topic name to broker config
+KAFKA_TOPIC_TO_CLUSTER: Mapping[str, Mapping[str, TopicDefinition | None]] = {
```
Comment (fpacifici) [inline src/sentry/conf/server.py:3465]: In snuba these are called `KAFKA_TOPIC_MAP` and `KAFKA_BROKER_CONFIG`. 
I think there would be value in consolidate the naming as well. It reduces the cognitive overhead when managing moving from one config (snuba) to the other (sentry).

Any chance we can name them in the same way? Even if this naming is more appropriate than the Snuba one.

File: src/sentry/conf/server.py [hunk 4]
```diff
@@ -3510,9 +3511,6 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
     KAFKA_SHARED_RESOURCES_USAGE: {"cluster": "default"},
 }
 
-
-# If True, consumers will create the topics if they don't exist
-KAFKA_CONSUMER_AUTO_CREATE_TOPICS = True
```
Comment (fpacifici) [inline src/sentry/conf/server.py:3515]: Why is this removed ?

File: src/sentry/conf/types/consumer_definition.py [hunk 5]
```diff
@@ -5,11 +5,17 @@
...
 class ConsumerDefinition(TypedDict, total=False):
     # Which logical topic from settings to use.
-    topic: Required[str | Callable[[], str]]
-    default_topic: str
+    topic: Topic
+
+    # Override topic. To be deprecated
+    topic_override: Required[str | Callable[[], str]]
+
+    validate_schema: str
```
Comment (fpacifici) [inline src/sentry/conf/types/consumer_definition.py:18]: Is this the name of the schema in the schema registry ?

File: src/sentry/consumers/__init__.py [hunk 6]
```diff
@@ -163,109 +164,122 @@ def ingest_monitors_options() -> list[click.Option]:
...
-# overridden. This is because the current topic override mechanism means the default topic name
-# is no longer available anywhere in code. We should probably fix this later so we don't need both
-#  "topic" and "default_topic" here though.
+# XXX: topic_override will be deprecated here in favor of KAFKA_TOPIC_OVERRIDE once all overrides in
+# prod have been migrated
 KAFKA_CONSUMERS: Mapping[str, ConsumerDefinition] = {
     "ingest-profiles": {
-        "topic": settings.KAFKA_PROFILES,
+        "topic": Topic.PROFILES,
+        "topic_override": settings.KAFKA_PROFILES,
```
Comment (fpacifici) [inline src/sentry/consumers/__init__.py:172]: Isn't this supposed to be replaced by the new `KAFKA_TOPIC_OVERRIDES` ?
I assume this is still referencing the old setting for backward compatibility during the rollout ?

File: src/sentry/utils/batching_kafka_consumer.py [hunk 7]
```diff
@@ -53,8 +52,6 @@ def create_topics(cluster_name: str, topics: list[str], force: bool = False) ->
 
     topics must be from the same cluster.
     """
-    if settings.KAFKA_CONSUMER_AUTO_CREATE_TOPICS or force:
-        conf = kafka_config.get_kafka_admin_cluster_options(cluster_name)
-        # Topics are implicitly created here
-        admin_client = AdminClient(conf)
-        wait_for_topics(admin_client, topics)
+    conf = kafka_config.get_kafka_admin_cluster_options(cluster_name)
```
Comment (fpacifici) [inline src/sentry/utils/batching_kafka_consumer.py:55]: Do we still have this consumer around ?

File: src/sentry/conf/server.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/conf/server.py:3515]: I cleaned it up since it doesn't do anything. Moved into https://github.com/getsentry/sentry/pull/65891

File: src/sentry/consumers/__init__.py (hunk 6 above)
Comment (untitaker) [reply src/sentry/consumers/__init__.py:172]: my interpretation is that we will gradually remove `topic_override` lines per-consumer, and right now this allows us to merge this PR before KAFKA_TOPIC_OVERRIDES is populated

it's confusing to me though that `KAFKA_TOPIC_OVERRIDES` is the new thing while `topic_override` is deprecated

**Review by untitaker — APPROVED** [review]

File: src/sentry/conf/server.py (hunk 3 above)
Comment (lynnagara) [reply src/sentry/conf/server.py:3465]: I intentionally named it differently because it is not the same as Snuba at all. Here we map the topic to a cluster definition. Snuba does not go through any cluster definition and all the topic configs are mapped to the broker config. I think we might want to align them in the future but I'm not yet sure which way is better, and I want to avoid making this huge change in this PR right now.

File: src/sentry/conf/types/consumer_definition.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/conf/types/consumer_definition.py:18]: this is a typo - it's supposed to be a bool

File: src/sentry/consumers/__init__.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/consumers/__init__.py:172]: Yeah, this is temporary thing that is going away, and aligns to how it works today (it was just named `topic` before). Though because of the consumers defined in getsentry there will need to be multiple even more temporary steps before this :(

File: src/sentry/conf/server.py [hunk 8]
```diff
@@ -3455,25 +3455,24 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
...
+# Deprecated
 KAFKA_TOPICS: Mapping[str, TopicDefinition | None] = {
     KAFKA_EVENTS: {"cluster": "default"},
     KAFKA_EVENTS_COMMIT_LOG: {"cluster": "default"},
     KAFKA_TRANSACTIONS: {"cluster": "default"},
     KAFKA_TRANSACTIONS_COMMIT_LOG: {"cluster": "default"},
     KAFKA_OUTCOMES: {"cluster": "default"},
-    # When OUTCOMES_BILLING is None, it inherits from OUTCOMES and does not
-    # create a separate producer. Check ``track_outcome`` for details.
-    KAFKA_OUTCOMES_BILLING: None,
```
Comment (lynnagara) [inline src/sentry/conf/server.py:3476]: Moved this part into https://github.com/getsentry/sentry/pull/65998 as it can be applied independently (and more safely) than the rest of this PR

File: src/sentry/conf/server.py [hunk 9]
```diff
@@ -3468,17 +3469,51 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
 
 # spans
 KAFKA_SNUBA_SPANS = "snuba-spans"
-
-KAFKA_SUBSCRIPTION_RESULT_TOPICS = {
```
Comment (lynnagara) [inline src/sentry/conf/server.py:3472]: Moved into https://github.com/getsentry/sentry/pull/66078

File: src/sentry/conf/types/consumer_definition.py [hunk 10]
```diff
@@ -5,11 +5,19 @@
 
 import click
 
+from sentry.conf.types.topic_definition import Topic
+
 
 class ConsumerDefinition(TypedDict, total=False):
-    # Which logical topic from settings to use.
-    topic: Required[str | Callable[[], str]]
-    default_topic: str
```
Comment (lynnagara) [inline src/sentry/conf/types/consumer_definition.py:12]: `default_topic` is replaced by `validate_schema` since it was only actually controlling whether schema validation was run or not

File: src/sentry/consumers/__init__.py [hunk 11]
```diff
@@ -394,12 +393,17 @@ def get_stream_processor(
         ) from e
 
     strategy_factory_cls = import_string(consumer_definition["strategy_factory"])
-    logical_topic = consumer_definition["topic"]
```
Comment (lynnagara) [inline src/sentry/consumers/__init__.py:397]: This variable was previously called logical topic, but that was misleading. It was always actually the physical topic which could be overridden via settings.

File: src/sentry/conf/server.py [hunk 12]
```diff
@@ -3468,8 +3469,52 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
 
 # spans
 KAFKA_SNUBA_SPANS = "snuba-spans"
+# END DEPRECATED SECTION
+
+
+# Mapping of default Kafka topic name to custom names
+KAFKA_TOPIC_OVERRIDES: Mapping[str, str] = {
+    "generic-metrics-subscription-results": KAFKA_GENERIC_METRICS_SUBSCRIPTIONS_RESULTS
```
Comment (lynnagara) [inline src/sentry/conf/server.py:3477]: Temporary so the override works for this topic which is the only one defined in the new-style. Later this will be removed when all overrides are defined in production

Comment (fpacifici) [reply src/sentry/conf/server.py:3483]: I am not following why we need to set up the override this way for one topic only. Could you please provide some more context ?

File: src/sentry/conf/server.py [hunk 13]
```diff
@@ -3471,8 +3472,60 @@ def build_cdc_postgres_init_db_volume(settings: Any) -> dict[str, dict[str, str]
...
+    # TODO: This is temporary while we migrate between the old and new way of defining overrides.
+    # To be removed once this is defined in prod, along with KAFKA_GENERIC_METRICS_SUBSCRIPTIONS_RESULTS
+    # variable which will no longer be needed
+    "generic-metrics-subscription-results": KAFKA_GENERIC_METRICS_SUBSCRIPTIONS_RESULTS
+}
+
+
+# Mapping of default Kafka topic name to cluster name
+# as per KAFKA_CLUSTERS.
+# This must be the default name that matches the topic
```
Comment (fpacifici) [inline src/sentry/conf/server.py:3489]: The default topic name, right ?

File: src/sentry/conf/types/kafka_definition.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+from collections.abc import Callable, Mapping, Sequence
+from enum import Enum
+from typing import Any, Required, TypedDict
+
+import click
+
+
+class Topic(Enum):
```
Comment (fpacifici) [inline src/sentry/conf/types/kafka_definition.py:10]: It would be useful to add a docstring here explaining that these are the logical topics. This is because here is where people add topics, this is being used by many.

File: src/sentry/conf/types/kafka_definition.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    TRANSACTIONS_COMMIT_LOG = "snuba-transactions-commit-log"
+    OUTCOMES = "outcomes"
+    OUTCOMES_BILLING = "outcomes-billing"
+    EVENTS_SUBSCRIPTIONS_RESULTS = "events-subscription-results"
+    TRANSACTIONS_SUBSCRIPTIONS_RESULTS = "transactions-subscription-results"
+    GENERIC_METRICS_SUBSCRIPTIONS_RESULTS = "generic-metrics-subscription-results"
+    SESSIONS_SUBSCRIPTIONS_RESULTS = "sessions-subscription-results"
+    METRICS_SUBSCRIPTIONS_RESULTS = "metrics-subscription-results"
+    INGEST_EVENTS = "ingest-events"
+    INGEST_EVENTS_DLQ = "ingest-events-dlq"
```
Comment (fpacifici) [inline src/sentry/conf/types/kafka_definition.py:23]: Wait, is this something you added today ?

File: src/sentry/conf/types/kafka_definition.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    SNUBA_SPANS = "snuba-spans"
+
+
+class ConsumerDefinition(TypedDict, total=False):
+
+    # XXX: Eventually only Topic will be accepted here.
+    # For backward compatibility with getsentry, we must also
+    # support the physical override topic name (str, Callable[str], str)
+    # while the migration is taking place
+    topic: Required[Topic | str | Callable[[], str]]
```
Comment (fpacifici) [inline src/sentry/conf/types/kafka_definition.py:50]: So does this mean that this will be the logical name only ? If yes where does the override happens ?

File: src/sentry/consumers/__init__.py [hunk 17]
```diff
@@ -237,8 +239,8 @@ def ingest_events_options() -> list[click.Option]:
         },
     },
     "generic-metrics-subscription-results": {
-        "topic": settings.KAFKA_GENERIC_METRICS_SUBSCRIPTIONS_RESULTS,
-        "default_topic": "generic-metrics-subscription-results",
+        "topic": Topic.GENERIC_METRICS_SUBSCRIPTIONS_RESULTS,
```
Comment (fpacifici) [inline src/sentry/consumers/__init__.py:242]: Is there any specific reason this is the only topic being set via the `Topic` enum ?

File: src/sentry/consumers/__init__.py [hunk 18]
```diff
@@ -394,12 +396,19 @@ def get_stream_processor(
...
-    logical_topic = consumer_definition["topic"]
-    if not isinstance(logical_topic, str):
-        logical_topic = logical_topic()
+    consumer_topic = consumer_definition["topic"]
+    if isinstance(consumer_topic, Topic):
+        default_topic = consumer_topic.value
+        real_topic = settings.KAFKA_TOPIC_OVERRIDES.get(default_topic, default_topic)
+    else:
+        # TODO: Deprecated, remove once this way is no longer used
+        if not isinstance(consumer_topic, str):
```
Comment (fpacifici) [inline src/sentry/consumers/__init__.py:405]: This condition seems counterintuitive. Was that any reason not to phrase it positively as:
```
if isinstance(consumer_topic, str):
    real_topic = consumer_topic
else:
   ....
```

File: tests/sentry/conf/test_kafka_definitions.py [hunk 19]
```diff
@@ -1,10 +1,24 @@
...
+from sentry.conf.types.kafka_definition import (
+    ConsumerDefinition,
+    Topic,
+    validate_consumer_definition,
+)
 from sentry.consumers import KAFKA_CONSUMERS
 from sentry.testutils.cases import TestCase
 
 
+def test_topic_definition() -> None:
```
Comment (fpacifici) [inline tests/sentry/conf/test_kafka_definitions.py:13]: I think we should also have a test to verify that all the topics in `KAFKA_TOPIC_OVERRIDES` are actually in the `Topic` enum.

**Review by fpacifici — COMMENTED** [review]
Remembering how complicated this migration was in Snuba it will be critical to review the rollout plan. What is your strategy and which steps will you go through?
Also how are we going to deal with self hosted configuration ? As people may be using the old override system, we should make it clear it is deprecated and it is going to go away. In Snuba, if I remember correctly, we had loud warning log message claiming the config was deprecated.

File: src/sentry/conf/server.py (hunk 12 above)
Comment (lynnagara) [reply src/sentry/conf/server.py:3483]: Two reasons
- demonstration of how this works
- it is the only topic that has schema validation turned on. In order to keep that happening, we need to know both the default and override topic there so it must use the new system.

File: src/sentry/conf/types/kafka_definition.py (hunk 15 above)
Comment (lynnagara) [reply src/sentry/conf/types/kafka_definition.py:23]: yes

File: src/sentry/conf/types/kafka_definition.py (hunk 16 above)
Comment (lynnagara) [reply src/sentry/conf/types/kafka_definition.py:50]: override happens in settings and nowhere else, it will be fully removed from consumer definitions

File: src/sentry/consumers/__init__.py (hunk 17 above)
Comment (lynnagara) [reply src/sentry/consumers/__init__.py:242]: I'll give the same answer as above - two reasons:
- demonstration of how this works
- it is the only topic that has schema validation turned on. In order to keep that happening, we need to know both the default and override topic there so it must use the new system.

File: src/sentry/consumers/__init__.py (hunk 18 above)
Comment (lynnagara) [reply src/sentry/consumers/__init__.py:405]: I did not write this code, I copied it from line 398 above, and don't want to refactor it unnecessarily given it will probably go away in a couple of days.

File: src/sentry/conf/types/kafka_definition.py (hunk 14 above)
Comment (lynnagara) [reply src/sentry/conf/types/kafka_definition.py:10]: Added docstring. There is also a test now that asserts that these same values are present in sentry-kafka-schemas

File: src/sentry/conf/server.py (hunk 13 above)
Comment (lynnagara) [reply src/sentry/conf/server.py:3489]: yes

File: tests/sentry/conf/test_kafka_definitions.py (hunk 19 above)
Comment (lynnagara) [reply tests/sentry/conf/test_kafka_definitions.py]: This is always empty/default in dev/ci so it won't do much

**Comment by lynnagara** [comment]
> Remembering how complicated this migration was in Snuba it will be critical to review the rollout plan. What is your strategy and which steps will you go through? Also how are we going to deal with self hosted configuration ? As people may be using the old override system, we should make it clear it is deprecated and it is going to go away. In Snuba, if I remember correctly, we had loud warning log message claiming the config was deprecated.

Believe it or not I actually don't think this is going to be a particularly complicated or long migration. There is nothing to do in self-hosted since we don't really provide/document any way to to override topic configs. Though technically possible, I highly doubt anyone has figured out how to set up overrides and synchronize them across our codebases.

The next step will be to populate all the clusters and overrides in getsentry. I will do it one environment at a time to be safe, probably starting with the single tenants..

Once all the configuration is populated via both the old-way and new-way, then we can switch all the code paths to stop referencing the old variables, and after that delete them.

File: tests/sentry/conf/test_kafka_definitions.py (hunk 19 above)
Comment (fpacifici) [reply tests/sentry/conf/test_kafka_definitions.py]: oh right, I guess this is more for getsentry

**Comment by fpacifici** [comment]
> There is nothing to do in self-hosted since we don't really provide/document any way to to override topic configs. Though technically possible, I highly doubt anyone has figured out how to set up overrides and synchronize them across our codebases.

We do provide a way to override sentry settings and it is in the public docs. That contains topics config. In Snuba we did take care of self hosted as we were used to receive github issues asking to allow the kafka settings customization (like authentication) which means people did customize the kafka config.

I'd really consider adding a simple log warning (after we migrated prod) to state the settings are deprecated. Then some time later remove them.

File: src/sentry/conf/server.py (hunk 12 above)
Comment (fpacifici) [reply src/sentry/conf/server.py:3483]: ok

**Review by fpacifici — APPROVED** [review]
